mitmproxy/protocol/http.py

```diff
@@ -170,13 +172,33 @@ def __call__(self):
             try:
                 flow = models.HTTPFlow(self.client_conn, self.server_conn, live=self)
                 flow.request = request
+
                 # set upstream auth
                 if self.mode == "upstream" and self.config.upstream_auth is not None:
                     flow.request.headers["Proxy-Authorization"] = self.config.upstream_auth
                 self.process_request_hook(flow)
 
+                # WebSockets
+                if websockets.check_handshake(request.headers):
```

Comment:
I think this logic should sit at https://github.com/mitmproxy/mitmproxy/blob/cacee3871c6a9f0be7127f3c790e09a1daaf8490/mitmproxy/protocol/http.py#L193 - we don't know yet if the remote actually speaks websocket. Having it down there also enables us to fall back to TCP mode for versions != 13.